### DISCUSSION TITLE
Street Name Announcement Next Steps

### DISCUSSION THREAD

**Discussion by zyphlar** [organicmaps/organicmaps#8412] [Ideas]
Collecting ideas for the next revision of PR #3130.

- Don't call them exits unless they're motorways (maybe even not unless they have exit numbers, to avoid "exit onto" when the current road is trunk or lower, which feels weird. Likewise motorway splits can be odd when they say "exit onto London")
- More dynamic phrasing, like "in 500 feet, take exit 123 onto M-123 towards London" instead of always "exit onto M-123 London" or "take exit 123, M-123, London"
  - Using either more detailed formatter strings or an improved localization library, since conjugation and grammar becomes very complex
  - Support calling out left-hand (or right-hand in left-driving countries) exits to reduce confusion
- Consider lane guidance like "use the left two lanes to..."
- Use SSML IPA tags and name pronunciation OSM tags to pronounce difficult names, for supported TTS engines (Google Play is the only one I've found on Android, supposedly iPhone does and Amazon, but this needs testing)
- See if there is some technique to pronounce foreign words in the user's desired voice, like how a bilingual person might
- Use intersection names instead of (or alongside?) street names in certain countries
  - Japan
  - Some South American countries?
  - others?
- Enable by default in countries that expect it
  - North America 
  - others?

Feel free to add more ideas below, which can turn into issues or PRs in the future. Ideally each code change will be as small as possible for rapid development, testing, and release, while keeping an eye towards the eventual goals. Maintainers also want to avoid external dependencies as much as possible.

Related:
- Serbian grammar improvements: https://github.com/orgs/organicmaps/discussions/9702

**Comment by biodranik** [comment]
Any help with some existing direction issues would also be appreciated. E.g. adding missing directions, or removing excess directions.

**Reply by zyphlar** [reply]
Can you elaborate with examples?

**Reply by biodranik** [reply]
https://github.com/organicmaps/organicmaps/issues?q=is%3Aopen+is%3Aissue+label%3ADirections

**Reply by zyphlar** [reply]
Ah that's probably going to be changing the generation of "turns" in the router moreso than changing what is said at each turn. I'm sure I'll get into that but I haven't touched it yet

**Comment by RicoElectrico** [comment]
I tested the tts street name announcement and it seems to work well, even if it's erring on the verbose side. One problem might be that street names are spelled in full. Like "Kapitana Antoniego Ledóchowskiego" instead of just "Ledóchowskiego"  😅

That's how it is in OSM, not sure how to approach it.

https://github.com/balrog-kun/shrtnms
This repo could be a start.

Which also ties into the search, one would also benefit from down-weighing titles and given names.

**Reply by Urvish79** [reply]
Idea seems pretty interesting to work on!!!

**Reply by biodranik** [reply]
@RicoElectrico what about adding a special OSM tag on how to pronounce a street or place name in navigation systems?

**Comment by dvdmrtnz** [comment]
In my opinion, now that street names are announced, simplifying roundabout instructions is very important: https://github.com/organicmaps/organicmaps/issues/6146

**Comment by zyphlar** [comment]
@biodranik says:

For Greece street names using English TTS on iOS pronounces street names, while using Russian TTS doesn't. It sounds like "turn left on" instead of "turn left on XXX street".

It would be great to make a list of such cases for different TTS engines, and either disable street names in such cases, or try to transliterate street names (I didn't test ru + en instead of ru + greek).

**Comment by zyphlar** [comment]
@biodranik reports that some roads have refs which are displayed and announced unnecessarily. He suggests collecting the list of all ref values and see what can be filtered out. Strings starting with ru:regional look like a good candidate.

**Reply by biodranik** [reply]
Here are dumps of all ref values from the OSM planet. Raw dump, and dump with unique combinations, sorted by name and by count.

[raw_refs_sorted.txt.zip](https://github.com/user-attachments/files/16326160/raw_refs_sorted.txt.zip)
[refs_counted.txt.zip](https://github.com/user-attachments/files/16326166/refs_counted.txt.zip)
[refs_counted_sorted.txt.zip](https://github.com/user-attachments/files/16326169/refs_counted_sorted.txt.zip)

**Comment by Petak73** [comment]
As a side note - announcing street names always seemed to me like the least useful feature of a nav app. If you are in an area where the street names mean anything to you, you normally don't need navigation anyway.

**Reply by zyphlar** [reply]
This is a highly regional assumption. In Japan, street names aren't shown on intersections hardly at all, but rather major intersections prominently have crossing names, like "Meiji Crossing" etc.

I've discovered that in Eastern Europe, many street signs are so small they're impossible to read from a car or absent entirely, and the roads seem hierarchical enough that wayfinding with "turn left soon" is enough since there will only be a few major reasonable options.

However in USA, street names are prominently displayed and the only way to know what "in 500 feet, turn right" means, because humans aren't tape measures and in that range there could be five driveways, three alleys, and one or two major named roads. (Look up "stroad" for the completely chaotic way most of America is built.) It's absolutely crucial to announce street names in the USA or you can get lost or even into an accident. It's even more important for motorway exits and junctions: before this project, it would simply say "in 1 mile, exit." Which comically-if-not-tragically underestimates the number of exits an American interstate can pack into a mile.

Since many of this project's contributors and users are in Eastern Europe and countries that don't use street names to navigate, the feature remains off by default, but please don't tell someone (me) that the work they've done for months isn't useful, especially if you aren't aware of the context they're operating in. It's much safer and more useful to share *your* context, like for example the countries you've been to where street names aren't useful. Eventually we may automatically turn off/on the feature for select countries, since a GPS that doesn't have voice announcements with names is "broken" in the USA but "annoying" in Ukraine.

**Reply by Petak73** [reply]
You seem annoyed by my opinion and I can see why.
Still, telling me to refrain from expressing it just because you've spent time is ... funny :-)
I've been throughout Eastern, Central and Southern Europe and can't think of a single place where street names would normally be visible from far enough to be a meaningful automotive navigation hint.
Obviously, this is a non-issue if announcing street names is optional. Preferably not automatically selected based on region, but definitely optional.
Have a nice day,
Peter
PS I am yet to hear a street name announcement that doesn't sound ridiculous for any names other than those common in the English language.

**Reply by patepelo** [reply]
In Barcelona this feature is super useful
